### PR TITLE
Make the miner less likely to get stuck

### PR DESCRIPTION
**Pull request by Mekle001**
# Changes proposed in this pull request:
- Make getWorkingNode more stable/deterministic
- Add tracking and validation of rotation 
- Make node picking sequential until there are 10 nodes built on a level - this should do at least 2 nodes out from shaft on a new level

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java [hunk 1]
```diff
@@ -194,7 +195,7 @@ public Level(@NotNull final CompoundNBT compound)
     public Node getRandomNode(@Nullable final Node node)
     {
         Node nextNode = null;
-        if (node != null && rand.nextInt(RANDOM_TYPES) > 0)
+        if (node != null && getNumberOfBuiltNodes() > 10 && rand.nextInt(RANDOM_TYPES) > 0)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java]: constant please

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java [hunk 2]
```diff
@@ -230,6 +231,9 @@ public void closeNextNode(final int rotation, final Node node)
                 nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, ROTATE_ONCE));
                 nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, ROTATE_THREE_TIMES));
                 break;
+            case UNDEFINED:
+                Log.getLogger().warn("Node: " + node.getX() + ":" + node.getZ() +" style undefined, not creating children");
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java]: I'd say .error -> Please tell the mod authors about this.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java [hunk 3]
```diff
@@ -245,8 +249,14 @@ public void closeNextNode(final int rotation, final Node node)
             nodes.put(pos, tempNodeToAdd);
             openNodes.add(tempNodeToAdd);
         }
-        nodes.get(new Vec2i(tempNode.getX(), tempNode.getZ())).setStatus(Node.NodeStatus.COMPLETED);
+        Node I = nodes.get(new Vec2i(tempNode.getX(), tempNode.getZ()));
+        if(!tempNode.equals(I))
+        {
+            Log.getLogger().warn("Nodes not equal during close");
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java]: same here potentially

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java [hunk 4]
```diff
@@ -127,9 +139,18 @@ public static Node createFromNBT(@NotNull final CompoundNBT compound)
 
         //Set the node status in all directions.
         @NotNull final Node node = new Node(x, z, parent);
+        if(style == NodeType.UNDEFINED)
+        {
+            com.ldtteam.blockout.Log.getLogger().warn("Node " + x + "," + z + " has an undefined style");
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java]: wrong logger

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java [hunk 5]
```diff
@@ -367,6 +393,18 @@ public int hashCode()
...
+        //New node, undefined
+        UNDEFINED
+    }
+
+    public Optional<Integer> getRot()
+    {
+        return rot;
+    }
+
+    public void setRot(int rot)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java:408]: javadoc

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/5688/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

**Review by someaddons — APPROVED** [review]

**Pull request merged**